design.md

```diff
@@ -0,0 +1,138 @@
+# Kernel Design
+
+## Functional Requirements
+
+### gRPC Server
+
+The kernel runs a gRPC server that supports the following operations.
+
+#### Running Scripts
+
+- Operations:
+  - Run a script from its contents.
+  - Run a script from a Git File URI.
+- Dependency resolution:
+  - Scripts may import from other scripts in the project. These dependencies are resolved to the relevant files in the project.
+  - Scripts may depend on Bowler libraries specified by a Git File URI.
+    - By default, resolution uses GitFS to pull the script.
+    - If the URI resolves to a Bowler library that has been dev'd, then the dev'd version must be used instead of the version from the remote.
+  - If credentials are required to resolve a dependency, the kernel must try to load them from the local environment first. If that fails, the kernel must ask the IDE for authentication. The authentication received from the IDE must not be stored on disk. If no IDE is available, dependency resolution fails and an exception is thrown at the call site.
+- A script may be given arguments via a list of objects when it is invoked.
+- A script may return an object when it returns.
+- ! RFC: How are these objects serialized?
+
+#### GitFS
+
+- Each repo is uniquely identified by a Git Repo URI.
+- Each file in a repo is uniquely identified by a Git File URI.
+- Each Git URI includes a remote.
+- Operations:
+  - Clone or pull a repo.
+  - List files in a repo.
+  - Read the contents of a file in a repo.
+  - Run a script in a repo.
+    - May pass arguments as a list of objects.
+    - May receive a return value as an object.
+  - Load a robot from a robot config file in a repo.
+
+#### Hardware Management
+
+- Operations:
+  - Load a robot from its config file.
+    - May specify the contents of the config file.
+    - May specify a Git File URI.
+    - Must specify whether to connect to hardware or a simulator.
+      - In the hardware case, a connection method must be specified.
+      - In the simulator case, a simulator plugin must be specified.
```

Comment:
Question: is there a meaningful difference between "connection method" and "simulator plugin"? i.e. what's the difference between the interfaces for each? If there isn't one, then can they be the same thing?

Reply:
I have changed the wording a bit but the difference is that a "connection method" is a real thing right now. It specifies how the pc connects to the device (e.g. UDP, HID). We also will need to support simulator plugins and the kernel will need to know what it has to connect to.